packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +531,820 @@ class ShellRoute extends RouteBase {
+  /// absolute minimum** to avoid any unnecessary resource use.
+  final bool preload;
+
+  /// Restoration ID to save and restore the state of the navigator, including
+  /// its history.
+  final String? restorationScopeId;
+
+  /// The observers for this branch.
+  ///
+  /// The observers parameter is used by the [Navigator] built for this branch.
+  final List<NavigatorObserver>? observers;
+}
+
+/// The snapshot of the current state of a [StatefulShellRoute].
+///
+/// Note that this an immutable class, that represents the snapshot of the state
+/// of a StatefulShellRoute at a given point in time. Therefore, instances of
+/// this object should not be cached, but instead passed down from the builder
+/// functions of StatefulShellRoute.
+@immutable
+abstract class StatefulShellRouteState {
+  /// The associated [StatefulShellRoute]
+  StatefulShellRoute get route;
+
+  /// The GoRouterState associate with [route].
+  GoRouterState get routerState;
+
+  /// The index of the currently active [StatefulShellBranch].
+  ///
+  /// Corresponds to the index of the branch in the List returned from
+  /// branchBuilder of [StatefulShellRoute].
+  int get currentIndex;
+
+  /// The Navigator key of the current navigator.
+  GlobalKey<NavigatorState> get currentNavigatorKey;
+
+  /// Navigate to the current location of the shell navigator with the provided
+  /// index.
+  ///
+  /// This method will switch the currently active [Navigator] for the
+  /// [StatefulShellRoute] by replacing the current navigation stack with the
+  /// one of the route branch identified by the provided index. If resetLocation
+  /// is true, the branch will be reset to its initial location
+  /// (see [StatefulShellBranch.initialLocation]).
+  void goBranch({
+    required int index,
+  });
+}
+
+/// Builder for a custom container for shell route Navigators.
+typedef ShellNavigatorContainerBuilder = Widget Function(BuildContext context,
+    StatefulShellRouteState shellRouteState, List<Widget> children);
+
+/// Widget for managing the state of a [StatefulShellRoute].
+///
+/// Normally, this widget is not used directly, but is instead created
+/// internally by StatefulShellRoute. However, if a custom container for the
+/// branch Navigators is required, StatefulShellRouteController can be used in
+/// the builder or pageBuilder methods of StatefulShellRoute to facilitate this.
+/// The container is created using the provided [ShellNavigatorContainerBuilder],
+/// where the List of Widgets represent the Navigators for each branch.
+///
+/// Example:
+/// ```
+/// builder: (BuildContext context, StatefulShellRouteState state, Widget child) {
+///   return StatefulShellRouteController(
+///     shellRouteState: state,
+///     containerBuilder: (_, __, List<Widget> children) => MyCustomShell(shellState: state, children: children),
+///   );
+/// }
+/// ```
+abstract class StatefulShellRouteController extends StatefulWidget {
```

Comment:
StatefulShellRouteController is a confusing name. it makes me think it is similar TextEditingController or ScrollController, but it is not. on the other hand, I think we might as well just make _StatefulNavigationShell a public class.

Reply:
Yes, I debated with myself a bit about the name. Although I do think it can be seen as a controller for a StatefulShellRoute, since it manages the state and behaviour for it. But I'm ok with just using the name StatefulNavigationShell (or if we find an even better name).